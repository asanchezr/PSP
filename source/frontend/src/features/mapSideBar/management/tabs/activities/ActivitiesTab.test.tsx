
import Claims from '@/constants/claims';

import { cleanup, getMockRepositoryObj, render, RenderOptions, waitForEffects } from '@/utils/test-utils';
import ActivitiesTab, { IActivitiesTabProps } from './ActivitiesTab';
import { mockManagementFileResponse } from '@/mocks/managementFiles.mock';
import { ApiGen_CodeTypes_ManagementFileStatusTypes } from '@/models/api/generated/ApiGen_CodeTypes_ManagementFileStatusTypes';
import { toTypeCode } from '@/utils/formUtils';

const mockManagementFileApi = mockManagementFileResponse();

const mockGetManagementActivitiesApi = {
  error: undefined,
  response: undefined,
  execute: vi.fn(),
  loading: false,
};

vi.mock('@/hooks/repositories/useManagementActivityRepository', () => ({
  useManagementActivityRepository: () => {
    return {
      getManagementFileActivities: mockGetManagementActivitiesApi,
      getManagementActivities: mockGetManagementActivitiesApi,
      deleteManagementActivity: getMockRepositoryObj(),
    };
  },
}));


describe('ManagementSummaryView component', () => {
  // render component under test
  const setup = async (
    renderOptions: RenderOptions & { props?: Partial<IActivitiesTabProps> },
  ) => {
    const utils = render(
      <ActivitiesTab managementFile={renderOptions?.props?.managementFile ?? mockManagementFileApi}      />,
      {
        ...renderOptions,
        useMockAuthentication: true,
      },
    );

    return { ...utils };
  };

  beforeEach(() => {
    mockGetManagementActivitiesApi.execute.mockResolvedValue([]);
  });

  afterEach(() => {
    cleanup();
    vi.clearAllMocks();
  });

  it('matches snapshot', async () => {
    const { asFragment } = await setup({});
    await waitForEffects();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the edit button for users with management edit permissions', async () => {
    const { queryByTestId } = await setup({ claims: [Claims.MANAGEMENT_EDIT] });
    await waitForEffects();

    const editButton = queryByTestId('add-activity-button');
    expect(editButton).toBeInTheDocument();
  });

  it('does not render the edit button for users that do not have management edit permissions', async () => {
    const { queryByTestId } = await setup({
      claims: [],
    });
    await waitForEffects();

    const editButton = queryByTestId('add-activity-button');
    expect(editButton).not.toBeInTheDocument();
  });

  it.each([
    ['Management File Status is "Completed"', ApiGen_CodeTypes_ManagementFileStatusTypes.COMPLETE],
    ['Management File Status is "Archived"', ApiGen_CodeTypes_ManagementFileStatusTypes.ARCHIVED],
    ['Management File Status is "Hold"', ApiGen_CodeTypes_ManagementFileStatusTypes.HOLD],
    ['Management File Status is "Cancelled"', ApiGen_CodeTypes_ManagementFileStatusTypes.CANCELLED],
  ])(
    'renders the warning icon for management files in non-editable status - %s',
    async (_: string, fileStatus: ApiGen_CodeTypes_ManagementFileStatusTypes) => {
      const mockManagementFile = {
        ...mockManagementFileResponse(),
        fileStatusTypeCode: toTypeCode(fileStatus),
      };

      const { queryByTestId } = await setup({
        props: {
          managementFile: mockManagementFile,
        },
        claims: [Claims.MANAGEMENT_EDIT],
      });
      await waitForEffects();

      const icon = queryByTestId('tooltip-icon-1-summary-cannot-edit-tooltip');
      expect(icon).toBeInTheDocument();
    },
  );
});
