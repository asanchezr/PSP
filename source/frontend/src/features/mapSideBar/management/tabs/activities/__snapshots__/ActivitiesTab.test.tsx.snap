// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ManagementSummaryView component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div />
  .c7 {
  float: right;
  cursor: pointer;
  font-size: 3.2rem;
}

.c0 {
  background-color: #f2f2f2;
  padding-top: 1rem;
}

.c2 {
  font-weight: bold;
  color: var(--theme-blue-100);
  border-bottom: 0.2rem var(--theme-blue-90) solid;
  margin-bottom: 2.4rem;
}

.c1 {
  margin: 1.6rem;
  padding: 1.6rem;
  background-color: white;
  text-align: left;
  border-radius: 0.5rem;
}

.c3 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  min-height: 4.5rem;
}

.c3 .btn {
  margin: 0;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-left: 0.5rem;
}

.c6 {
  width: 1.6rem;
  height: 1.6rem;
}

.c5 {
  width: 1.6rem;
  height: 1.6rem;
}

<div
    class="c0"
  >
    <div
      class="c1 form-section"
    >
      <h2
        class="c2"
      >
        <div
          class="no-gutters row"
        >
          <div
            class="align-content-end col"
          >
            <div
              class="c3 no-gutters row"
            >
              <div
                class="d-flex justify-content-start align-items-end col-auto"
              >
                Activities List
              </div>
              <div
                class="d-flex justify-content-end my-1 col-auto"
              />
            </div>
          </div>
        </div>
      </h2>
      <div
        class="collapse show"
      >
        <p>
          <strong>
            Activity documentation:
          </strong>
           You can attach a document after creating the activity. Create, then edit and attach a file if needed.
        </p>
        <div
          class="table"
          data-testid="mgmt-activity-list"
          role="table"
        >
          <div
            class="thead thead-light"
          >
            <div
              class="tr"
              role="row"
              style="display: flex; flex: 1 0 auto; min-width: 185px;"
            >
              <div
                class="th"
                colspan="1"
                role="columnheader"
                style="box-sizing: border-box; flex: 35 0 auto; min-width: 35px; width: 35px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
              >
                <div
                  class="sortable-column"
                >
                  Activity type
                  <div
                    class="c4"
                    data-testid="sort-column-activityType"
                  >
                    <svg
                      class="c5"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      style="margin-bottom: -8px;"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 320l128-128 128 128z"
                      />
                    </svg>
                    <svg
                      class="c6"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 192l128 128 128-128z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="th"
                colspan="1"
                role="columnheader"
                style="box-sizing: border-box; flex: 35 0 auto; min-width: 35px; width: 35px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
              >
                <div
                  class="sortable-column"
                >
                  Activity sub-type
                  <div
                    class="c4"
                    data-testid="sort-column-activitySubTypes"
                  >
                    <svg
                      class="c5"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      style="margin-bottom: -8px;"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 320l128-128 128 128z"
                      />
                    </svg>
                    <svg
                      class="c6"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 192l128 128 128-128z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="th"
                colspan="1"
                role="columnheader"
                style="box-sizing: border-box; flex: 35 0 auto; min-width: 35px; width: 35px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
              >
                <div
                  class="sortable-column"
                >
                  Activity status
                  <div
                    class="c4"
                    data-testid="sort-column-activityStatusType"
                  >
                    <svg
                      class="c5"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      style="margin-bottom: -8px;"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 320l128-128 128 128z"
                      />
                    </svg>
                    <svg
                      class="c6"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 192l128 128 128-128z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="th"
                colspan="1"
                role="columnheader"
                style="box-sizing: border-box; flex: 50 0 auto; min-width: 50px; width: 50px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
              >
                <div
                  class="sortable-column"
                >
                  Commencement
                  <div
                    class="c4"
                    data-testid="sort-column-requestedAddedDate"
                  >
                    <svg
                      class="c5"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      style="margin-bottom: -8px;"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 320l128-128 128 128z"
                      />
                    </svg>
                    <svg
                      class="c6"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 192l128 128 128-128z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="th"
                colspan="1"
                role="columnheader"
                style="box-sizing: border-box; flex: 20 0 auto; min-width: 30px; width: 20px; justify-content: center; text-align: center; flex-wrap: wrap; align-items: center; display: flex;"
              >
                <div
                  class="sortable-column"
                >
                  Actions
                </div>
              </div>
            </div>
          </div>
          <div
            class="no-rows-message"
          >
            No property management activities found
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="align-self-center col-auto"
          />
          <div
            class="ml-auto align-self-center col-auto"
          />
          <div
            class="align-self-center col-auto"
          />
        </div>
      </div>
    </div>
    <div
      class="c1 form-section"
    >
      <h2
        class="c2"
      >
        <div
          class="no-gutters row"
        >
          <div
            class="align-content-end col"
          >
            <div
              class="d-flex"
            >
              <div
                class="no-gutters justify-content-between align-items-end mr-2 row"
              >
                <div
                  class="my-1 col-auto"
                >
                  Ad-hoc Activities List
                </div>
              </div>
              <span
                class="tooltip-icon align-self-end"
                data-testid="tooltip-icon-property-file-activity-summary"
                id="property-file-activity-summary"
              >
                <svg
                  class="tooltip-icon"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            class="pl-2 d-flex align-items-end col-auto"
          >
            <svg
              class="c7"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>
                collapse-section
              </title>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="m7 14 5-5 5 5z"
              />
            </svg>
          </div>
        </div>
      </h2>
      <div
        class="collapse show"
      >
        <div
          class="table"
          data-testid="adhoc-activity-list-readonly"
          role="table"
        >
          <div
            class="thead thead-light"
          >
            <div
              class="tr"
              role="row"
              style="display: flex; flex: 1 0 auto; min-width: 185px;"
            >
              <div
                class="th"
                colspan="1"
                role="columnheader"
                style="box-sizing: border-box; flex: 35 0 auto; min-width: 35px; width: 35px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
              >
                <div
                  class="sortable-column"
                >
                  Activity type
                  <div
                    class="c4"
                    data-testid="sort-column-activityType"
                  >
                    <svg
                      class="c5"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      style="margin-bottom: -8px;"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 320l128-128 128 128z"
                      />
                    </svg>
                    <svg
                      class="c6"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 192l128 128 128-128z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="th"
                colspan="1"
                role="columnheader"
                style="box-sizing: border-box; flex: 35 0 auto; min-width: 35px; width: 35px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
              >
                <div
                  class="sortable-column"
                >
                  Activity sub-type
                  <div
                    class="c4"
                    data-testid="sort-column-activitySubTypes"
                  >
                    <svg
                      class="c5"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      style="margin-bottom: -8px;"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 320l128-128 128 128z"
                      />
                    </svg>
                    <svg
                      class="c6"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 192l128 128 128-128z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="th"
                colspan="1"
                role="columnheader"
                style="box-sizing: border-box; flex: 35 0 auto; min-width: 35px; width: 35px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
              >
                <div
                  class="sortable-column"
                >
                  Activity status
                  <div
                    class="c4"
                    data-testid="sort-column-activityStatusType"
                  >
                    <svg
                      class="c5"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      style="margin-bottom: -8px;"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 320l128-128 128 128z"
                      />
                    </svg>
                    <svg
                      class="c6"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 192l128 128 128-128z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="th"
                colspan="1"
                role="columnheader"
                style="box-sizing: border-box; flex: 50 0 auto; min-width: 50px; width: 50px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
              >
                <div
                  class="sortable-column"
                >
                  Commencement
                  <div
                    class="c4"
                    data-testid="sort-column-requestedAddedDate"
                  >
                    <svg
                      class="c5"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      style="margin-bottom: -8px;"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 320l128-128 128 128z"
                      />
                    </svg>
                    <svg
                      class="c6"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128 192l128 128 128-128z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="th"
                colspan="1"
                role="columnheader"
                style="box-sizing: border-box; flex: 35 0 auto; min-width: 30px; width: 35px; justify-content: center; text-align: center; flex-wrap: wrap; align-items: center; display: flex;"
              >
                <div
                  class="sortable-column"
                >
                  Navigation
                </div>
              </div>
            </div>
          </div>
          <div
            class="no-rows-message"
          >
            No property management activities found
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="align-self-center col-auto"
          />
          <div
            class="ml-auto align-self-center col-auto"
          />
          <div
            class="align-self-center col-auto"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
